enhancements/network/multi-networkpolicy.md

```diff
@@ -0,0 +1,207 @@
+   changes.
+
+The `Metadata` section above is intended to support the creation of tooling
+around the enhancement process.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This proposal to bring networkpolicy features for multus interface.
+
+## Motivation
+
+Customer wants to make their Kubernetes secure, by filtering traffic from
+unexpected source. Kubernetes provides `NetworkPolicy` feature, however, it
+is only applied to default network interface, not additional network interface
+that is created by Multus CNI. NetworkPolicy for Multus Interface brings
+NetworkPolicy mechanism to additional network interface and it provides customer
+to filter traffic from unexpected source.
+
+### Goals
+
+- Introduce MultiNetworkPolicy CRD
+- Introduce multi-networkpolicy which generates iptables from above CRD
+- Provide iptable based filtering mechanism for pod
+- Gathering related information (iptables rules in pod network namespace) by must-gather
+
+### Non-Goals
+
+- Provide eBPF filtering mechanism
+- Provide NetworkPolicy for userspace pod which uses DPDK and so on
+
+## Proposal
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Story 1
+
+Customer can specify how groups of pods are allowed to communicate with each other
+and other network endpoints as Kubernetes NetworkPolicy does.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Multus
+
+No additional change is required because multi-networkpolicy only consumes 
+network-attachment-definition CRD and it does not change net-attach-def CRD.
+
+### How user enables multi-networkpolicy?
+
+Currently we're going to bring it as Tech Preview first and then GA after a few release,
+hence opt-in style configuration is planned at Tech Preview and make it default at GA.
+Opt-in style configuration will be done at cluster-network-operator as boolean flag
+(such as `useMultiNetworkPolicy: true`).
+
+
+### Risks and Mitigations
+
+## Design Details
+
+### MultiNetworkPolicy samples
+
+MultiNetworkPolicy CRD has pretty similar schema to Kubernetes NetworkPolicy and its
+semantics is same as well as Kubernetes other than target interface. Target interface
+is specified annotation, `k8s.v1.cni.cncf.io/policy-for`, as net-attach-def name.
+
+```
+apiVersion: k8s.cni.cncf.io/v1beta1
+kind: MultiNetworkPolicy
+metadata:
+  name: policy-test1
+  annotations:
+    k8s.v1.cni.cncf.io/policy-for: macvlan-net1
```

Comment:
Is there any use-case for having `*/policy-for` multiple net-attach-defs? E.g. can you specify:

```
k8s.v1.cni.cncf.io/policy-for: macvlan-net1, foo-net, bar-net, quux-net
```

Reply:
Good question. I don't get real customer requirements yet, however, I believe they have, for example, network isolation between traffic planes. In this case, customer may want to apply same policy between different net-attach-def. But it depends on case and there's pros/cons. It's quite controversial.

Do you have any idea for that?

Reply:
Is it an option to do this the other way around, for example, refers to the MultiNetworkPolicy name in the net-attach-def annotation (defines a new annotation that points to the MultiNetworkPolicy name) and the namespace isolation can also be achieved by checking if the net-attach-def and its associated MultiNetworkPolicy are in the same namespace. The good thing for this approach is that one interface (net-attach-def) can only map to one Policy and multiple net-attach-defs can use the same Policy (no need to create/update Policy for newly created net-attach-def).

Reply:
This is a trade-off. In your case, one net-attach-def, of course, can only map to one policy, but it need to change Pod annotation. Usually, many annotation in the Pod is created when the Pod is launched (net-attach-def annotation, netowork annotation and so on). So if we put policy annotation in pod, it might be against other Pod yaml.

On the other side, K8s NetworkPolicy yaml is used to be changed on the fly. When the user changed networkpolicy then K8s network policy is changed. So Many user suppose that NetworkPolicy object can be changed any time. So I wish that adding annotation in MultiNetworkPolicy is more feasible than adding it in Pod.

Reply:
Tomo I didn't mean to change pod annotation, but adding a new policy annotation in net-attach-def so that every net-attach-def is associated with a network policy.

Reply:
I see. But net-attach-def is regarded as 'static' object, once we create it then we keep it and not changed every time.